src/Servers/Kestrel/Core/src/Internal/Http3/Http3ControlStream.cs

```diff
@@ -289,7 +292,12 @@ private ValueTask ProcessSettingsFrameAsync(ReadOnlySequence<byte> payload)
         }
 
         _haveReceivedSettingsFrame = true;
-        using var closedRegistration = _context.StreamContext.ConnectionClosed.Register(state => ((Http3ControlStream)state!).OnStreamClosed(), this);
+        _completeFeature.OnCompleted(state =>
+        {
+            ((Http3ControlStream)state!).OnStreamClosed();
+            _connectionClosed = true;
```

Comment:
This might currently allocate?
```suggestion
            ((Http3ControlStream)state!)._connectionClosed = true;
```
Maybe make delegate static?